pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -174,19 +214,163 @@ func (w *HashAggFinalWorker) receiveFinalResultHolder() (*chunk.Chunk, bool) {
+				(*restoreadData)[key] = results
+
+				for aggPos := 0; aggPos < aggFuncNum; aggPos++ {
+					results[aggPos] = partialResultsRestored[aggPos][rowPos]
+				}
+			}
+		}
+	}
+	return totalMemDelta, nil
+}
+
+func (w *HashAggFinalWorker) restoreOnePartition(ctx sessionctx.Context) (bool, error) {
+	restoredData := make(aggfuncs.AggPartialResultMapper)
+	bInMap := 0
+
+	restoredPartitionIdx, isSuccess := w.spillHelper.getNextPartition()
+	if !isSuccess {
+		return false, nil
+	}
+
+	spilledFilesIO := w.spillHelper.getListInDisks(restoredPartitionIdx)
+	for _, spilledFile := range spilledFilesIO {
+		memDelta, err := w.restoreFromOneSpillFile(ctx, &restoredData, spilledFile, &bInMap)
+		if err != nil {
+			return false, err
+		}
+
+		w.memTracker.Consume(memDelta)
+	}
+	w.spilledDataChan <- &restoredData
+	return true, nil
+}
+
+func (w *HashAggFinalWorker) mergeResultsAndSend(ctx sessionctx.Context) (bool, error) {
+	spillContinue, err := w.consumeIntermData(ctx)
+	if err != nil {
+		w.outputCh <- &AfFinalResult{err: err}
+		return false, err
+	}
+
+	w.loadFinalResult(ctx)
+	return spillContinue, nil
+}
+
 func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	w.spilledDataChan = make(chan *aggfuncs.AggPartialResultMapper, 1)
 	start := time.Now()
 	defer func() {
 		if r := recover(); r != nil {
 			recoveryHashAgg(w.outputCh, r)
 		}
+		close(w.spilledDataChan)
 		if w.stats != nil {
 			w.stats.WorkerTime += int64(time.Since(start))
 		}
 		waitGroup.Done()
 	}()
-	if err := w.consumeIntermData(ctx); err != nil {
-		w.outputCh <- &AfFinalResult{err: err}
+
+	enableIntest := false
+	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
+		if val.(bool) {
+			enableIntest = true
+		}
+	})
+
+	if intest.InTest && enableIntest {
+		num := rand.Intn(50)
+		if num == 0 {
+			panic("Intest panic: final worker is panicked before start")
+		} else if num == 1 {
+			time.Sleep(2 * time.Millisecond)
+		}
+	}
+
+	w.isSpilledTriggered = w.spillHelper.isSpillTriggered()
+	if w.isSpilledTriggered {
+		if w.spillHelper.checkError() {
+			return
+		}
+
+		for {
+			hasData, err := w.restoreOnePartition(ctx)
+			if err != nil {
+				w.outputCh <- &AfFinalResult{err: err}
+				return
+			}
+
+			if !hasData {
+				return
+			}
+
+			if intest.InTest && enableIntest {
+				num := rand.Intn(10000)
+				if num < 7 {
+					panic("Intest panic: final worker is panicked when running")
+				} else if num < 14 {
+					time.Sleep(2 * time.Millisecond)
+				} else if num < 21 {
+					w.memTracker.Consume(1000000)
+				} else if num < 28 {
+					w.outputCh <- &AfFinalResult{err: errors.Errorf("Random fail is triggered in final worker")}
+					return
+				}
+			}
+
+			var spillContinue bool
+			spillContinue, err = w.mergeResultsAndSend(ctx)
+			if err != nil || !spillContinue {
+				return
+			}
+		}
+	} else {
+		_, err := w.mergeResultsAndSend(ctx)
+		if err != nil {
+			return
+		}
 	}
-	w.loadFinalResult(ctx)
 }
```

Comment:
``` suggestion
func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
	w.spilledDataChan = make(chan *aggfuncs.AggPartialResultMapper, 1)
	start := time.Now()
	defer w.cleanup(start, waitGroup)

	failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
		if val.(bool) {
			w.intestBeforeStart()
		}
	})

	w.isSpilledTriggered = w.spillHelper.isSpillTriggered()
	if w.isSpilledTriggered {
		w.handleSpilledData(ctx)
	} else {
		_, err := w.mergeResultsAndSend(ctx)
		if err != nil {
			return
		}	
	}
}

func (w *HashAggFinalWorker) cleanup(start time.Time, waitGroup *sync.WaitGroup) {
	if r := recover(); r != nil {
		recoveryHashAgg(w.outputCh, r)
	}
	close(w.spilledDataChan)
	if w.stats != nil {
		w.stats.WorkerTime += int64(time.Since(start))
	}
	waitGroup.Done()
}

func (w *HashAggFinalWorker) handleSpilledData(ctx sessionctx.Context) {
	if w.spillHelper.checkError() {
		return
	}

	for {
		hasData, err := w.restoreOnePartition(ctx)
		if err != nil {
			w.outputCh <- &AfFinalResult{err: err}
			return
		}
		if !hasData {
			return
		}
		failpoint.Inject("enableAggSpillIntest", func(val failpoint.Value) {
			if val.(bool) {
				w.intestDuringRun()
			}
		})
		spillContinue, err := w.mergeResultsAndSend(ctx)
		if err != nil || !spillContinue {
			return
		}
	}
}

func (w *HashAggFinalWorker) intestBeforeStart() {
    num := rand.Intn(50)
    if num == 0 {
        panic("Intest panic: final worker is panicked before start")
    } else if num == 1 {
        time.Sleep(2 * time.Millisecond)
    }
}

func (w *HashAggFinalWorker) intestDuringRun() {
    num := rand.Intn(10000)
    if num < 7 {
        panic("Intest panic: final worker is panicked when running")
    } else if num < 14 {
        time.Sleep(2 * time.Millisecond)
    } else if num < 21 {
        w.memTracker.Consume(1000000)
    } else if num < 28 {
        w.outputCh <- &AfFinalResult{err: errors.Errorf("Random fail is triggered in final worker")}
    }
}
```

Reply:
Done